File: SDK Examples/Examples/ANSI_C/Counter_BufferedPulseWidthMeasurement/BufferedPulseWidthMeasurement.c
/*******************************************************************************
Copyright (c) 1983-2017 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 
================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log: $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*     BufferedPulseWidthMeasurement.c
*
* Example Category:
*    Counter
*
* Description:
*    This example demonstrates how to use Buffered Pulse Width Measurement function.
*
* Instructions for Running:
*    1  Set the 'deviceDescription' for opening the device. 
*	  2  Set the 'profilePath' to save the profile path of being initialized device. 
*
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include "../inc/compatibility.h"
#include "../../../inc/bdaqctrl.h"

//-----------------------------------------------------------------------------------
// Configure the following parameters before running the demo
//-----------------------------------------------------------------------------------
#define     deviceDescription L"PCIE-1813,BID#15"
const wchar_t* profilePath = L"../../profile/PCIE-1813.xml";
#define     sectionLength 1024
#define     sectionCount  0

// user buffer size should be equal or greater than raw data buffer length, because data ready count
// is equal or more than smallest section of raw data buffer and up to raw data buffer length.
// users can set 'USER_BUFFER_SIZE' according to demand.
#define    USER_BUFFER_SIZE   sectionLength
PulseWidth userDataBuffer[USER_BUFFER_SIZE];

void waitAnyKey()
{
   do{SLEEP(1);} while(!kbhit());
} 

void BDAQCALL OnDataReadyEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{	
   ErrorCode ret = Success;
	int32 returnedCount = 0;
	BufferedPwMeterCtrl *ctrl = (BufferedPwMeterCtrl *)sender;
   int32 getDataCount = min(USER_BUFFER_SIZE, args->Count); 
	ret = BufferedPwMeterCtrl_GetData(ctrl, args->Channel, getDataCount, userDataBuffer, 0, &returnedCount, NULL);
	// in this demo, we show only the first sample of each channel's new data
   if(!BioFailed(ret)) {
      printf("Buffered PulseWidth Measurement get data count is  %d\n", returnedCount);
      printf("the first sample for Channel %d is: Pulse Width: HiPeriod=%f s  LoPeriod=%f s\n", args->Channel, userDataBuffer[0].HiPeriod, userDataBuffer[0].LoPeriod);
   }
}
// This function is used to deal with 'Overrun' Event.
void BDAQCALL OnOverRunEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
	printf("Buffered PulseWidth Measurement Overrun: channel = %d, offset = %d, count = %d\n", args->Channel, args->Offset, args->Count);
}
// This function is used to deal with 'CacheOverflow' Event.
void BDAQCALL OnCacheOverflowEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
	printf("Buffered PulseWidth Measurement Cache Overflow: channel = %d, offset = %d, count = %d\n", args->Channel, args->Offset, args->Count);
}
// This function is used to deal with 'Stopped' Event.
void BDAQCALL OnStoppedEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
   printf("Buffered PulseWidth Measurement stopped: channel = %d, offset = %d, count = %d\n", args->Channel, args->Offset, args->Count);
}

int main(int argc, char* argv[])
{
   ErrorCode ret = Success;
   int32     chCount     = 4;
   int8      chEnabled[] = {1, 0, 0, 0};
   Trigger*  trigger     = NULL;
   Record*   record      = NULL;

   // Step 1: Create a 'BufferedPwMeterCtrl' for Buffered PulseWidth Measurement function.
   BufferedPwMeterCtrl* bfdPwMeterCtrl = BufferedPwMeterCtrl_Create();

   // Step 2: Set the notification event Handler by which we can known the state of operation effectively.
   BufferedPwMeterCtrl_addDataReadyHandler(bfdPwMeterCtrl, OnDataReadyEvent, NULL);
   BufferedPwMeterCtrl_addOverrunHandler(bfdPwMeterCtrl, OnOverRunEvent, NULL);
   BufferedPwMeterCtrl_addCacheOverflowHandler(bfdPwMeterCtrl, OnCacheOverflowEvent, NULL);
   BufferedPwMeterCtrl_addStoppedHandler(bfdPwMeterCtrl, OnStoppedEvent, NULL);
   do
   {
      // Step 3: Select a device by device number or device description and specify the access mode.
      // in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
      DeviceInformation devInfo;
      devInfo.DeviceNumber = -1;
      devInfo.DeviceMode   = ModeWrite;
      devInfo.ModuleIndex  = 0;
      wcscpy(devInfo.Description, deviceDescription);
      ret = BufferedPwMeterCtrl_setSelectedDevice(bfdPwMeterCtrl, &devInfo);
      CHK_RESULT(ret);
      ret = BufferedPwMeterCtrl_LoadProfile(bfdPwMeterCtrl, profilePath);//Loads a profile to initialize the device.
      CHK_RESULT(ret);

      // Step 4: Set necessary parameters.
      ret = BufferedPwMeterCtrl_setChannelMap(bfdPwMeterCtrl, chCount, chEnabled);
	   CHK_RESULT(ret);

      ret = BufferedPwMeterCtrl_setSampleClkSrc(bfdPwMeterCtrl, SignalNone);
      CHK_RESULT(ret);

      record = BufferedPwMeterCtrl_getRecord(bfdPwMeterCtrl);
      {
         ret = Record_setSectionCount(record, sectionCount); //The 0 means setting 'streaming' mode
         CHK_RESULT(ret);
         ret = Record_setSectionLength(record, sectionLength);
         CHK_RESULT(ret);
      }

      trigger = BufferedPwMeterCtrl_getTrigger(bfdPwMeterCtrl); //Setting trigger
      if(trigger) {
         ret = Trigger_setSource(trigger, SignalNone); // The "SignalNone" means doing without trigger.
         CHK_RESULT(ret);
      }

      // Step 5: Start Buffered PulseWidth Measurement Counter 
      ret = BufferedPwMeterCtrl_Prepare(bfdPwMeterCtrl);
      CHK_RESULT(ret);

	   ret= BufferedPwMeterCtrl_setEnabled(bfdPwMeterCtrl, TRUE);
	   CHK_RESULT(ret);

	   // Step 6: The device is acquiring PulseWidth data.
	   printf("Pulse Width Measurement is in progress... any key to quit !\n");
	   while ( !kbhit()) {
         SLEEP(1);
      }

	   // step 7: Stop the operation if it is running.
	  ret= BufferedPwMeterCtrl_setEnabled(bfdPwMeterCtrl, FALSE);
     CHK_RESULT(ret);
   }while(FALSE);

   // Step 8: Close device, release any allocated resource.
	BufferedPwMeterCtrl_Dispose(bfdPwMeterCtrl);

	// If something wrong in this execution, print the error code on screen for tracking.
   if(BioFailed(ret))
   {
      printf("\nSome error occurred. And the last error code is 0x%X.\n", ret);
      waitAnyKey();
   }  
   return 0;
}

File: SDK Examples/Examples/Matlab/Counter_BufferedPulseWidthMeasurement/BufferedPulseWidthMeasurement.m
% Buffered PulseWidthMeasurement.m
%
% Example Category:
%    Counter
% Matlab(2010 or 2010 above)
%
% Description:
%    This example demonstrates how to use Buffered Pulse Width Measurement function.
%
% Instructions for Running:
%    1. Set the 'deviceDescription' for opening the device. 
%    2. Set the 'chEnable' to select channel of the counter to operate.
%    3. Set the 'sectionLength'to decide what occasion to signal event; 
%    4. Set the 'sectionCount' to decide the count of data section to 
%       acquire,the operation will stop automatically after the specified 
%       count of data section is got.  
%    5. Set the 'sampleClkSrc' to set sampling clock source.
%
% I/O Connections Overview:
%    Please refer to your hardware reference manual.

function BufferedPulseWidthMeasurement()

% Make Automation.BDaq assembly visible to MATLAB.
BDaq = NET.addAssembly('Automation.BDaq4');

% Configure the following three parameters before running the demo.
% The default device of project is PCIE-1813, users can choose other 
% devices according to their needs.
% General settings
deviceDescription = 'PCIE-1813,BID#15';
chEnable = [1,0,0,0]; 
sectionLength = int32(1024); 
sectionCount = int32(0); % "sectionCount" is zero, meaning Streaming Mode.
sampleClkSrc = Automation.BDaq.SignalDrop.SignalNone;

% Triggrt settings
% Set trigger parameters.
% Remove comment marker if use Trigger.
% triggerAction = Automation.BDaq.TriggerAction.DelayToStart;
triggerSource = Automation.BDaq.SignalDrop.SignalNone;
% triggerEdge = Automation.BDaq.ActiveSignal.RisingEdge;
% triggerLevel = double(3.0);

% Step 1: Create a 'BufferedPwMeterCtrl' for Buffered Pulse Width Measurement function.
bfdPwMterCtrl = Automation.BDaq.BufferedPwMeterCtrl();

% Step 2: Set the notification event Handler by which we can known the 
% state of operation effectively. 
addlistener(bfdPwMterCtrl, 'DataReady', @bfdPwMterCtrl_DataReady);

try
    % Step 3: Select a device by device number or device description and 
    % specify the access mode. In this example we use 
    % ModeWrite(default) mode so that we can fully control the 
    % device, including configuring, sampling, etc.
    bfdPwMterCtrl.SelectedDevice = Automation.BDaq.DeviceInformation(...
        deviceDescription);
    
    if 	bfdPwMterCtrl.Features.TriggerCount
    else
        e = MException('DAQError:NotSupport', ...
            'The device do not support trigger function!');
        throw (e);
    end
    
    % Step 4: Set necessary parameters for counter operation.
    bfdPwMterCtrl.ChannelMap = chEnable;

    record = bfdPwMterCtrl.Record;
    record.SectionLength = sectionLength;
    record.SectionCount = sectionCount;
    bfdPwMterCtrl.SampleClkSrc = sampleClkSrc;
    
    % Setp 5: Trigger parameters setting.
    % Diffenert device with different trigger source, when changing trigger
    % source refer to manual for details.
    trigger = bfdPwMterCtrl.Trigger;
    if ~isempty(trigger)
        trigger.Source = triggerSource;
        %trigger.Action = triggerAction;
        %trigger.Edge = triggerEdge;
        %trigger.Level = triggerLevel;
    else
        e = MException('DAQError:NotSupport',...
            'The device do not support trigger function!');
        throw (e);
    end
    
    % Step 6: Start BufferedPulseWidthMeasurement
    errorCode = bfdPwMterCtrl.Prepare();
    if BioFailed(errorCode)
        throw Exception(); 
    end
    bfdPwMterCtrl.Enabled = true;
    
    % Step 7: Get Pulse Width value.
    fprintf('BufferedPulseWidthMeasurement is in progress...\n');
    fprintf('Connect the input signal to the connector.\n');
    input('Press Enter key to quit!\n\n','s');
 
    % Step 8: Stop BufferedPulseWidthMeasurement
    bfdPwMterCtrl.Enabled = false;
catch e
    % Something is wrong. 
    errStr = e.message;
    disp(errStr);
end   

% Step 9: Close device and release any allocated resource.
bfdPwMterCtrl.Dispose();

end

function result = BioFailed(errorCode)
result =  errorCode < Automation.BDaq.ErrorCode.Success && ...
    errorCode >= Automation.BDaq.ErrorCode.ErrorHandleNotValid;
end

function bfdPwMterCtrl_DataReady(sender, e)
bfdPwMterCtrl = sender;
dataBuf = NET.createArray('Automation.BDaq.PulseWidth', e.Count);
[errorCode, returnedCount, startTime] = bfdPwMterCtrl.GetData(e.Channel, e.Count, dataBuf, 0);
fprintf('BufferedPulseWidthMeasurement DataReadyEvent, channel %d  High Period:%f s, Low Period: %f s\n', ...
    e.Channel, dataBuf(1).HiPeriod, dataBuf(1).LoPeriod);
end

























